File: MMORPGServer/Proto/CreateLuaProto/Mail_Get_DetailProto.lua
--获取邮件详情
Mail_Get_DetailProto = { ProtoCode = 17002, IsSuccess = false, MailTitle = "", MailContent = "", ErrorMsg = "" }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
Mail_Get_DetailProto.__index = Mail_Get_DetailProto;

function Mail_Get_DetailProto.New()
    local self = { }; --初始化self
    setmetatable(self, Mail_Get_DetailProto); --将self的元表设定为Class
    return self;
end


--发送协议
function Mail_Get_DetailProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteBool(proto.IsSuccess);
    if(proto.IsSuccess) then
        ms:WriteUTF8String(MailTitle);
        ms:WriteUTF8String(MailContent);
        else
        ms:WriteUTF8String(ErrorMsg);
    end

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function Mail_Get_DetailProto.GetProto(buffer)

    local proto = Mail_Get_DetailProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.IsSuccess = ms:ReadBool();
    if(proto.IsSuccess) then
        proto.MailTitle = ms:ReadUTF8String();
        proto.MailContent = ms:ReadUTF8String();
        else
        proto.ErrorMsg = ms:ReadUTF8String();
    end

    ms:Dispose();
    return proto;
end

File: MMORPGServer/Proto/CreateProto/Mail_Get_DetailProto.cs
//===================================================
//作    者：崔炜斌
//创建时间：2019-12-13 19:29:52
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary>
/// 获取邮件详情
/// </summary>
public struct Mail_Get_DetailProto : IProto
{
    public ushort ProtoCode { get { return 17002; } }

    public bool IsSuccess; //是否成功
    public string MailTitle; //邮件标题
    public string MailContent; //邮件内容
    public string ErrorMsg; //错误信息

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteBool(IsSuccess);
            if(IsSuccess)
            {
                ms.WriteUTF8String(MailTitle);
                ms.WriteUTF8String(MailContent);
            }
            else
            {
                ms.WriteUTF8String(ErrorMsg);
            }
            return ms.ToArray();
        }
    }

    public static Mail_Get_DetailProto GetProto(byte[] buffer)
    {
        Mail_Get_DetailProto proto = new Mail_Get_DetailProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.IsSuccess = ms.ReadBool();
            if(proto.IsSuccess)
            {
                proto.MailTitle = ms.ReadUTF8String();
                proto.MailContent = ms.ReadUTF8String();
            }
            else
            {
                proto.ErrorMsg = ms.ReadUTF8String();
            }
        }
        return proto;
    }
}